### ISSUE TITLE
Exiting the REPL should be `.exit` instead of `exit`

### ISSUE THREAD

**Issue by kitsonk** [denoland/deno#4772]
Currently you exit the REPL via `exit`.  This is inconsistent with Node.js (which is `.exit`) and also collides with what is a valid variable name in JavaScript.

```
> const exit = "";
error: Uncaught SyntaxError: Identifier 'exit' has already been declared
► <unknown>:1:1
    at evaluate ($deno$/repl.ts:64:34)
    at Object.replLoop ($deno$/repl.ts:153:13)
> exit = "bar"
bar
> 
```

and even funnier/dangerous, this will cause the Deno REPL to exit:

```
> console.log(exit)
```

**Comment by ry** [comment]
It's just that isn't javascript. Can we just remove it instead?

I use ctrl+d to exit. Maybe we can just print a reminder for people when the repl starts?

**Comment by Caesar2011** [comment]
I use Ctrl+D all the time, much quicker and cooler. You can send the signal issued by Ctrl+d when using pipes, too.

Maybe we could also show the version number within the informational text on start.

**Comment by axetroy** [comment]
> I use Ctrl+D all the time, much quicker and cooler. You can send the signal issued by Ctrl+d when using pipes, too.
> 
> Maybe we could also show the version number within the informational text on start.

agree.

```
Welcome to Node.js v12.14.1.
Type ".help" for more information.
> .help
.break    Sometimes you get stuck, this gets you out
.clear    Alias for .break
.editor   Enter editor mode
.exit     Exit the repl
.help     Print this help message
.load     Load JS from a file into the REPL session
.save     Save all evaluated commands in this REPL session to a file

Press ^C to abort current expression, ^D to exit the repl
```

**Comment by kitsonk** [comment]
Only point, forcing CTRL-D only is an accessibility issue.  Being able to type something out without needing to hold another key down is always preferable from that aspect.

**Comment by Soremwar** [comment]
Let's make it a function then, like Python

`Deno.exit()` (?) Could work well

@kitsonk Is right however, typed commands provide better accessibility

Not sure if Deno is available in the REPL though

**Comment by Caesar2011** [comment]
I just checked and `Deno` is available in REPL.

**Comment by kitsonk** [comment]
As well as `window.close()` works.  I still think people would expect something terser than that.  What is the problem/concern with `.exit`?  In its syntax position it is never valid JavaScript.

**Comment by Soremwar** [comment]
>  In its syntax position it is never valid JavaScript.
That's the problem, you have to handle this cases in a different way you would handle JS code. That means more code to maintain

Why bother? Better to use already existing code
Functions make it clear that you will be execution and action also

Just my opinion though

**Comment by kitsonk** [comment]
> That's the problem, you have to handle this cases in a different way you would handle JS code.

Exactly, it is something that doesn't work in other environments.  `exit` is already handled differently, and very strangely.

**Comment by Caesar2011** [comment]
How about:

`Deno.exit()` is a bit longer than just `exit` or `.exit`. But it's a accessibility feature only and the shortcut should be preferred.

```
$~>deno
Welcome to deno 0.41.0. Press Ctrl+D or type "Deno.exit()" to exit REPL.
>
```

BTW: Does Deno has a help page in REPL or in general?

**Comment by Soremwar** [comment]
@Caesar2011 It doesn't

@kitsonk Well, what do you think? In the end it's all a matter of preference

`exit` could be patched to execute only when alone, so no `const exit or `console.log(exit)` oddities. But as I said, more code to maintain

**Comment by nayeemrmn** [comment]
@Caesar2011 It does, type `help`.

**Comment by Caesar2011** [comment]
> @Caesar2011 It does, type `help`.

The problem I see, and that's why I asked about the `help` command. If we want to implement other commands like help or save like node. With the same problem like now with exit. And in this case we should be consistent. `Deno.help()` and `Deno.save()` do not make any sense, but `.save` and `.help` does. As special REPL commands they do not have to be compliant with JS code. And `.exit` is in some way also a special command - or at least - expected if we have `.help`, `.load` and `.save`.

**Comment by kevinkassimo** [comment]
How about just using the `_` prefix? We already have `_` and `_error`, we can simply have `_exit` - as terse as `.exit` and is valid JS, and unlikely something people would name in REPL

**Comment by Soremwar** [comment]
@kevinkassimo I like that idea

**Comment by nayeemrmn** [comment]
No, `_`, `_error`, and `help`/`_help` work because they are just values, though automatically set. Exit is not the same, we should choose between `.exit` and removal (i.e. just using `Deno.exit()`).

Either way, if we support more REPL commands then we need a pattern for it, and `.<cmd>` is the only one that would generally make sense. Consider if we added a break command for example. These are meta REPL commands, they shouldn't be valid JS.

**Comment by Soremwar** [comment]
Is there a need for help inside `repl` though?

Why not just do `deno repl --help`?

My opinion, again, Repl should be limited to the execution of code, hence all the input it should take should be valid JavaScript code

**Comment by Soremwar** [comment]
Same with other arguments, 
```batch
load
save
help
```

Makes more sense to pass them as arguments **before** launching the Repo instead of inside of it

**Comment by nayeemrmn** [comment]
@Soremwar Doing everything before is not interactive. That's the whole point of a REPL. You don't start one with an exact plan. So yeah, there's no way to avoid non-JS commands if we want those features.

**Comment by Soremwar** [comment]
@nayeemrmn I don't know

Editor does seem like a nice feature Repl should have

But save seems a little insecure in this context. I don't know if this code is sandboxed as well, but piercing a whole in the sandbox to grant this one access to the filesystem seems a little compromising

**Comment by nayeemrmn** [comment]
> @nayeemrmn I don't know
> 
> `editor` does seem like a nice feature Repl should have
> 
> But `save` seems a little insecure in this context. I don't know if this code is sandboxed as well, but piercing a whole in the sandbox to grant this one access to the filesystem seems a little compromising

@Soremwar That's irrelevant, but the REPL has all permissions. I recommend playing around with it  a bit... you also didn't know earlier if it has access to the Deno namespace :p

**Comment by Soremwar** [comment]
@nayeemrmn To be fair, yeah XD

**Comment by gewoonwoutje** [comment]
Personally, I do use `exit`. I'm using it to close shell sessions as well. 
I'm all for removing it though, but a reminder like @Caesar2011 mentioned how to close the REPL would be nice.

**Comment by Fenzland** [comment]
Why not process the line REPL before pass it to V8 as JavaScript code, if the line matches `/^\s*exit\s*$/` then exit. So we can both keep the feature and prevent to add something to JS.

**Comment by kevinkassimo** [comment]
> Either way, if we support more REPL commands then we need a pattern for it, and `.<cmd>` is the only one that would generally make sense. Consider if we added a break command for example. These are meta REPL commands, they shouldn't be valid JS. (So what @Caesar2011 said.)

I guess I am not fully on board that we must use non-JS syntax to do meta commands that also makes sense in JS like `_exit` (since it make sense to be a JS thing as it could be implemented as a getter that have side effects), but I do agree for other ones beyond JS where there is room for debate. V8 uses `%` prefix for some operations so that would also be an interesting candidate.

> Why not process the line REPL before pass it to V8 as JavaScript code, if the line matches `/^\s*exit\s*$/` then exit. So we can both keep the feature and prevent to add something to JS.

Preprocessing is an option, though the regex you propose might be too broad.

**Comment by jsejcksn** [comment]
I think arbitrarily appropriating valid, top-level/global variables in order to use them as REPL environment commands will be an unwelcome surprise to some people.

**Comment by hayd** [comment]
We need to decide on this before 1.0.

Personally I think copying node is sensible/unsurprising/useful. Starting with a . is not js/ts so we can lookup the command and reject unknown commands without (prior to) evaluating. No need for any unexpected globals...

**Comment by kitsonk** [comment]
Closed via #5143
